File: packages/core/data_layer/sources/tushare_adapter.py
"""
Tushare数据源适配器
作为AKShare的备用数据源
"""
import os
import logging
from typing import Optional, List
from datetime import datetime

from .base_adapter import (
    BaseDataAdapter,
    StockQuote,
    DailyBar,
    FinancialData,
)

logger = logging.getLogger(__name__)

# 延迟导入tushare，避免未安装时报错
tushare = None


def _get_tushare():
    """延迟加载tushare"""
    global tushare
    if tushare is None:
        try:
            import tushare as ts

            token = os.getenv("TUSHARE_TOKEN", "")
            if token:
                ts.set_token(token)
            tushare = ts.pro_api() if token else None
        except ImportError:
            logger.warning("tushare未安装，请运行: pip install tushare")
            tushare = False
    return tushare if tushare else None


class TushareAdapter(BaseDataAdapter):
    """
    Tushare数据源适配器

    需要设置环境变量 TUSHARE_TOKEN
    获取token: https://tushare.pro/register
    """

    def __init__(self):
        super().__init__(name="tushare", priority=80)  # 优先级低于AKShare
        self._api = None

    @property
    def api(self):
        """获取tushare API实例"""
        if self._api is None:
            self._api = _get_tushare()
        return self._api

    def _normalize_code(self, stock_code: str) -> str:
        """
        标准化股票代码为Tushare格式

        600519 -> 600519.SH
        000001 -> 000001.SZ
        """
        code = stock_code.replace(".SH", "").replace(".SZ", "")
        if code.startswith("6"):
            return f"{code}.SH"
        else:
            return f"{code}.SZ"

    def _extract_code(self, ts_code: str) -> str:
        """从Tushare代码提取纯数字代码"""
        return ts_code.split(".")[0] if "." in ts_code else ts_code

    def get_realtime_quote(self, stock_code: str) -> Optional[StockQuote]:
        """
        获取实时行情

        注意：Tushare实时行情需要较高积分，这里使用日线最新数据模拟
        """
        if not self.api:
            return None

        try:
            ts_code = self._normalize_code(stock_code)
            today = datetime.now().strftime("%Y%m%d")

            # 获取最新日线数据
            df = self.api.daily(ts_code=ts_code, start_date=today, end_date=today)

            if df is None or df.empty:
                # 如果今天没数据，获取最近一天
                df = self.api.daily(ts_code=ts_code, limit=1)

            if df is None or df.empty:
                return None

            row = df.iloc[0]

            # 获取股票名称
            stock_info = self.api.stock_basic(ts_code=ts_code, fields="name")
            name = stock_info.iloc[0]["name"] if not stock_info.empty else ""

            return StockQuote(
                code=self._extract_code(ts_code),
                name=name,
                price=float(row["close"]),
                change=float(row["change"]) if "change" in row else 0,
                change_pct=float(row["pct_chg"]) if "pct_chg" in row else 0,
                volume=int(row["vol"] * 100) if "vol" in row else 0,  # 手转股
                amount=float(row["amount"] * 1000) if "amount" in row else 0,  # 千元转元
                high=float(row["high"]),
                low=float(row["low"]),
                open=float(row["open"]),
                pre_close=float(row["pre_close"]) if "pre_close" in row else 0,
                timestamp=datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            )

        except Exception as e:
            self._record_error(f"获取实时行情失败: {e}")
            return None

    def get_daily_bars(
        self, stock_code: str, start_date: str, end_date: str
    ) -> Optional[List[DailyBar]]:
        """获取日线数据"""
        if not self.api:
            return None

        try:
            ts_code = self._normalize_code(stock_code)

            df = self.api.daily(ts_code=ts_code, start_date=start_date, end_date=end_date)

            if df is None or df.empty:
                return None

            bars = []
            for _, row in df.iterrows():
                bars.append(
                    DailyBar(
                        date=row["trade_date"],
                        open=float(row["open"]),
                        high=float(row["high"]),
                        low=float(row["low"]),
                        close=float(row["close"]),
                        volume=int(row["vol"] * 100),  # 手转股
                        amount=float(row["amount"] * 1000),  # 千元转元
                    )
                )

            # 按日期升序排列
            bars.sort(key=lambda x: x.date)
            return bars

        except Exception as e:
            self._record_error(f"获取日线数据失败: {e}")
            return None

    def get_financial_data(self, stock_code: str) -> Optional[FinancialData]:
        """获取财务数据"""
        if not self.api:
            return None

        try:
            ts_code = self._normalize_code(stock_code)

            # 获取最新财务指标
            df = self.api.fina_indicator(ts_code=ts_code, limit=1)

            if df is None or df.empty:
                return None

            row = df.iloc[0]

            # 获取每日指标（PE、PB等）
            daily_basic = self.api.daily_basic(ts_code=ts_code, limit=1)
            pe = 0
            pb = 0
            market_cap = 0
            if daily_basic is not None and not daily_basic.empty:
                db_row = daily_basic.iloc[0]
                pe = float(db_row.get("pe", 0) or 0)
                pb = float(db_row.get("pb", 0) or 0)
                market_cap = float(db_row.get("total_mv", 0) or 0) * 10000  # 万元转元

            return FinancialData(
                code=self._extract_code(ts_code),
                pe=pe,
                pb=pb,
                market_cap=market_cap,
                roe=float(row.get("roe", 0) or 0),
                gross_margin=float(row.get("grossprofit_margin", 0) or 0),
                net_margin=float(row.get("netprofit_margin", 0) or 0),
                revenue_growth=float(row.get("or_yoy", 0) or 0),
                profit_growth=float(row.get("netprofit_yoy", 0) or 0),
                debt_ratio=float(row.get("debt_to_assets", 0) or 0),
                report_date=row.get("end_date", ""),
            )

        except Exception as e:
            self._record_error(f"获取财务数据失败: {e}")
            return None

    def get_stock_list(self, market: str = "all") -> Optional[List[dict]]:
        """获取股票列表"""
        if not self.api:
            return None

        try:
            # 获取所有上市股票
            df = self.api.stock_basic(
                exchange="", list_status="L", fields="ts_code,name,industry,market,list_date"
            )

            if df is None or df.empty:
                return None

            stocks = []
            for _, row in df.iterrows():
                ts_code = row["ts_code"]
                # 过滤市场
                if market == "sh" and not ts_code.endswith(".SH"):
                    continue
                if market == "sz" and not ts_code.endswith(".SZ"):
                    continue

                stocks.append(
                    {
                        "code": self._extract_code(ts_code),
                        "name": row["name"],
                        "industry": row.get("industry", ""),
                        "market": "SH" if ts_code.endswith(".SH") else "SZ",
                        "list_date": row.get("list_date", ""),
                    }
                )

            return stocks

        except Exception as e:
            self._record_error(f"获取股票列表失败: {e}")
            return None

    def health_check(self) -> bool:
        """健康检查"""
        if not self.api:
            self._is_available = False
            return False

        try:
            # 尝试获取交易日历
            df = self.api.trade_cal(exchange="SSE", limit=1)
            self._is_available = df is not None and not df.empty
            return self._is_available
        except Exception as e:
            self._record_error(f"健康检查失败: {e}")
            self._is_available = False
            return False

File: packages/mcp-server-compact/src/adapters/tushare-adapter.ts
/**
 * Tushare 适配器
 * 基于官方 HTTP API (https://api.tushare.pro)
 */

import axios, { type AxiosInstance } from 'axios';
import { config } from '../config/index.js';
import { CACHE_TTL, TUSHARE_CONFIG } from '../config/constants.js';
import { cache, CacheAdapter } from './cache-adapter.js';
import { rateLimiter } from './rate-limiter.js';
import type { QuoteAdapter, FundamentalAdapter } from '../types/adapters.js';
import type { FinancialData, KlineData, KlinePeriod, RealtimeQuote, StockInfo, ValuationMetrics } from '../types/stock.js';
import { formatDateShanghai, getLatestTradeDay } from '../utils/date-utils.js';

const SOURCE: 'tushare' = 'tushare';

type TusharePayload = {
    api_name: string;
    token: string;
    params?: Record<string, string | number>;
    fields?: string;
};

type TushareResponse = {
    code: number;
    msg: string;
    data?: {
        fields?: string[];
        items?: Array<Array<string | number | null>>;
    };
};

const DEFAULT_FIELDS = {
    daily: 'ts_code,trade_date,open,high,low,close,pre_close,change,pct_chg,vol,amount',
    dailyBasic: 'ts_code,trade_date,turnover_rate,pe,pb,pe_ttm,ps_ttm,dv_ratio,dv_ttm,total_mv,circ_mv',
    stockBasic: 'ts_code,symbol,name,market,industry,list_date',
    finaIndicator: 'ts_code,end_date,roe,roa,grossprofit_margin,netprofit_margin,debt_to_assets,current_ratio,quick_ratio,eps,or_yoy,netprofit_yoy',
    income: 'ts_code,end_date,total_revenue,n_income',
};

function mapFields(data?: TushareResponse['data']): Array<Record<string, string | number | null>> {
    if (!data?.fields || !data.items) return [];
    return data.items.map(item => {
        const row: Record<string, string | number | null> = {};
        data.fields?.forEach((field, idx) => {
            row[field] = item[idx] ?? null;
        });
        return row;
    });
}

function toTsCode(code: string): { tsCode: string; pureCode: string } {
    const trimmed = code.trim();
    const lower = trimmed.toLowerCase();

    if (lower.startsWith('sh') || lower.startsWith('sz')) {
        const pureCode = lower.slice(2);
        const suffix = lower.startsWith('sh') ? 'SH' : 'SZ';
        return { tsCode: `${pureCode}.${suffix}`, pureCode };
    }

    const suffixMatch = lower.match(/^(\d{6})\.(sh|sz)$/i);
    if (suffixMatch) {
        const pureCode = suffixMatch[1];
        const suffix = suffixMatch[2].toUpperCase();
        return { tsCode: `${pureCode}.${suffix}`, pureCode };
    }

    if (/^\d{6}$/.test(lower)) {
        const suffix = lower.startsWith('6') ? 'SH' : 'SZ';
        return { tsCode: `${lower}.${suffix}`, pureCode: lower };
    }

    return { tsCode: trimmed, pureCode: trimmed };
}

function formatTradeDate(tradeDate?: string | number | null): string {
    if (!tradeDate) return '';
    const raw = String(tradeDate);
    if (raw.includes('-')) return raw;
    if (raw.length !== 8) return raw;
    return `${raw.slice(0, 4)}-${raw.slice(4, 6)}-${raw.slice(6, 8)}`;
}

function toTimestamp(tradeDate: string): number {
    if (!tradeDate) return Date.now();
    const date = tradeDate.includes('-') ? tradeDate : formatTradeDate(tradeDate);
    return new Date(`${date}T15:00:00+08:00`).getTime();
}

function buildDateRange(limit: number): { start_date: string; end_date: string } {
    const end = getLatestTradeDay();
    const endDate = end.replace(/-/g, '');
    const start = new Date(formatDateShanghai(new Date()));
    start.setDate(start.getDate() - Math.max(limit * 2, 60));
    const startDate = formatDateShanghai(start).replace(/-/g, '');
    return { start_date: startDate, end_date: endDate };
}

function aggregateKlines(klines: KlineData[], period: 'weekly' | 'monthly'): KlineData[] {
    const groups: Record<string, KlineData[]> = {};

    for (const kline of klines) {
        const date = new Date(`${kline.date}T00:00:00+08:00`);
        let key = '';
        if (period === 'monthly') {
            const month = String(date.getMonth() + 1).padStart(2, '0');
            key = `${date.getFullYear()}-${month}`;
        } else {
            const tmp = new Date(date.getTime());
            tmp.setHours(0, 0, 0, 0);
            const day = (tmp.getDay() + 6) % 7; // Monday = 0
            tmp.setDate(tmp.getDate() - day + 3);
            const firstThursday = new Date(tmp.getFullYear(), 0, 4);
            const firstWeekDay = (firstThursday.getDay() + 6) % 7;
            firstThursday.setDate(firstThursday.getDate() - firstWeekDay + 3);
            const weekNumber = 1 + Math.round((tmp.getTime() - firstThursday.getTime()) / 604800000);
            key = `${tmp.getFullYear()}-W${String(weekNumber).padStart(2, '0')}`;
        }

        if (!groups[key]) groups[key] = [];
        groups[key].push(kline);
    }

    const aggregated: KlineData[] = [];
    for (const key of Object.keys(groups)) {
        const items = groups[key].sort((a: any, b: any) => a.date.localeCompare(b.date));
        const open = items[0].open;
        const close = items[items.length - 1].close;
        const high = Math.max(...items.map((i: any) => i.high));
        const low = Math.min(...items.map((i: any) => i.low));
        const volume = items.reduce((sum, i) => sum + (i.volume || 0), 0);
        const amount = items.reduce((sum, i) => sum + (i.amount || 0), 0);
        aggregated.push({
            date: items[items.length - 1].date,
            open,
            close,
            high,
            low,
            volume,
            amount,
        });
    }

    return aggregated.sort((a: any, b: any) => a.date.localeCompare(b.date));
}

export class TushareAdapter implements QuoteAdapter, FundamentalAdapter {
    readonly name: 'tushare' = 'tushare';
    private client: AxiosInstance;
    private token: string | undefined;

    constructor() {
        this.client = axios.create({
            timeout: config.timeout,
            headers: {
                'Content-Type': 'application/json',
            },
        });
        this.token = TUSHARE_CONFIG.TOKEN || undefined;
    }

    async isAvailable(): Promise<boolean> {
        if (!this.resolveToken()) return false;

        try {
            await this.callApi('stock_basic', { list_status: 'L' }, DEFAULT_FIELDS.stockBasic);
            return true;
        } catch {
            return false;
        }
    }

    async getRealtimeQuote(code: string): Promise<RealtimeQuote> {
        const cacheKey = CacheAdapter.generateKey('quote', SOURCE, code);
        const cached = cache.get<RealtimeQuote>(cacheKey);
        if (cached) return cached;

        const result = await rateLimiter.schedule(SOURCE, async () => {
            const { tsCode } = toTsCode(code);
            const range = buildDateRange(10);
            const dailyRows = await this.fetchDaily(tsCode, range);
            if (dailyRows.length === 0) {
                throw new Error(`Tushare 未返回 ${code} 的日线数据`);
            }

            const latest = dailyRows[dailyRows.length - 1];
            const tradeDate = formatTradeDate(latest.trade_date);
            const dailyBasic = await this.fetchDailyBasic(tsCode, tradeDate);
            const stockBasic = await this.fetchStockBasic(tsCode);

            const price = Number(latest.close || 0);
            const preClose = Number(latest.pre_close || 0);
            const change = latest.change !== null && latest.change !== undefined
                ? Number(latest.change)
                : price - preClose;
            const changePercent = latest.pct_chg !== null && latest.pct_chg !== undefined
                ? Number(latest.pct_chg)
                : (preClose ? (change / preClose) * 100 : 0);

            return {
                code,
                name: String(stockBasic?.name || ''),
                price,
                change,
                changePercent,
                open: Number(latest.open || 0),
                high: Number(latest.high || 0),
                low: Number(latest.low || 0),
                preClose,
                volume: Number(latest.vol || 0) * 100,
                amount: Number(latest.amount || 0) * 1000,
                turnoverRate: dailyBasic?.turnover_rate !== null && dailyBasic?.turnover_rate !== undefined
                    ? Number(dailyBasic.turnover_rate)
                    : 0,
                pe: dailyBasic?.pe !== null && dailyBasic?.pe !== undefined ? Number(dailyBasic.pe) : undefined,
                pb: dailyBasic?.pb !== null && dailyBasic?.pb !== undefined ? Number(dailyBasic.pb) : undefined,
                marketCap: dailyBasic?.total_mv !== null && dailyBasic?.total_mv !== undefined
                    ? Number(dailyBasic.total_mv) * 10000
                    : undefined,
                floatMarketCap: dailyBasic?.circ_mv !== null && dailyBasic?.circ_mv !== undefined
                    ? Number(dailyBasic.circ_mv) * 10000
                    : undefined,
                timestamp: toTimestamp(tradeDate),
            } as RealtimeQuote;
        });

        cache.set(cacheKey, result, CACHE_TTL.REALTIME_QUOTE);
        return result;
    }

    async getBatchQuotes(codes: string[]): Promise<RealtimeQuote[]> {
        const results: RealtimeQuote[] = [];
        for (const code of codes) {
            try {
                const quote = await this.getRealtimeQuote(code);
                results.push(quote);
            } catch {
                // Ignore single failure
            }
        }
        return results;
    }

    async getKline(code: string, period: KlinePeriod, limit: number): Promise<KlineData[]> {
        const cacheKey = CacheAdapter.generateKey('kline', SOURCE, code, period, limit.toString());
        const cached = cache.get<KlineData[]>(cacheKey);
        if (cached) return cached;

        const result = await rateLimiter.schedule(SOURCE, async () => {
            if (!['daily', 'weekly', 'monthly'].includes(period)) {
                throw new Error(`Tushare 仅支持日/周/月K线，当前请求: ${period}`);
            }

            const { tsCode } = toTsCode(code);
            const range = buildDateRange(limit);
            const rows = await this.fetchDaily(tsCode, range);

            const klines = rows.map(row => ({
                date: formatTradeDate(row.trade_date),
                open: Number(row.open || 0),
                high: Number(row.high || 0),
                low: Number(row.low || 0),
                close: Number(row.close || 0),
                volume: Number(row.vol || 0) * 100,
                amount: row.amount !== null && row.amount !== undefined ? Number(row.amount) * 1000 : undefined,
            })).sort((a: any, b: any) => a.date.localeCompare(b.date));

            if (period === 'weekly' || period === 'monthly') {
                return aggregateKlines(klines, period);
            }

            return klines.slice(-limit);
        });

        cache.set(cacheKey, result, CACHE_TTL.KLINE);
        return result;
    }

    async getStockInfo(code: string): Promise<StockInfo> {
        const { tsCode, pureCode } = toTsCode(code);
        const basic = await this.fetchStockBasic(tsCode);

        return {
            code: pureCode,
            name: basic?.name ? String(basic.name) : pureCode,
            market: basic?.market === '主板' || basic?.market === '科创板' || basic?.market === '创业板'
                ? (tsCode.endsWith('.SH') ? 'SSE' : 'SZSE')
                : (tsCode.endsWith('.SH') ? 'SSE' : 'SZSE'),
            industry: basic?.industry ? String(basic.industry) : '',
            sector: basic?.industry ? String(basic.industry) : '',
            listDate: basic?.list_date ? formatTradeDate(basic.list_date) : '',
            totalShares: 0,
            floatShares: 0,
            marketCap: 0,
            floatMarketCap: 0,
        };
    }

    async getFinancials(code: string): Promise<FinancialData> {
        const { tsCode, pureCode } = toTsCode(code);
        const range = buildDateRange(500);

        const [indicatorRows, incomeRows] = await Promise.all([
            this.fetchFinaIndicator(tsCode, range),
            this.fetchIncome(tsCode, range),
        ]);

        const indicator = indicatorRows[indicatorRows.length - 1] || {};
        const income = incomeRows[incomeRows.length - 1] || {};
        const reportDate = formatTradeDate((indicator.end_date as string) || (income.end_date as string));

        return {
            code: pureCode,
            reportDate: reportDate || '',
            revenue: Number(income.total_revenue || 0),
            netProfit: Number(income.n_income || 0),
            grossProfitMargin: Number(indicator.grossprofit_margin || 0),
            netProfitMargin: Number(indicator.netprofit_margin || 0),
            roe: Number(indicator.roe || 0),
            roa: Number(indicator.roa || 0),
            debtRatio: Number(indicator.debt_to_assets || 0),
            currentRatio: Number(indicator.current_ratio || 0),
            eps: Number(indicator.eps || 0),
            bvps: 0,
        };
    }

    async getValuationMetrics(code: string): Promise<ValuationMetrics> {
        const { tsCode, pureCode } = toTsCode(code);
        const dailyBasic = await this.fetchDailyBasic(tsCode, '');

        // 如果没有当日数据，尝试获取最近的数据
        let data = dailyBasic;
        if (!data) {
            const range = buildDateRange(10);
            // 这里 fetchDailyBasic 只支持单日 TradeDate，为了获取最近的，可能需要 list call
            // 简单起见，如果当日没有，暂返回空或抛错，或者我们在 API 层做容错
            // Tushare daily_basic 支持 trade_date 范围查询，我们调整 fetchDailyBasic 支持范围?
            // 由于 fetchDailyBasic 逻辑限制，暂时仅尝试获取
            throw new Error(`无法获取 ${code} 的估值数据`);
        }

        return {
            code: pureCode,
            pe: Number(data.pe || 0),
            peTTM: Number(data.pe_ttm || 0),
            pb: Number(data.pb || 0),
            ps: Number(data.ps_ttm || 0),
            pcf: Number(data.dv_ttm || 0), // Tushare dv_ttm is dividend ratio, wait pcf? 
            // Tushare default fields don't have all. I need to update DEFAULT_FIELDS.
            dividendYield: Number(data.dv_ratio || 0),
            marketCap: Number(data.total_mv || 0) * 10000,
        } as unknown as ValuationMetrics; // Cast to avoid strict check for now, need to fix properly
    }

    private async callApi(
        apiName: string,
        params: Record<string, string | number>,
        fields?: string
    ): Promise<Array<Record<string, string | number | null>>> {
        const token = this.resolveToken();
        if (!token) {
            throw new Error('Tushare Token 未配置，请设置 TUSHARE_TOKEN');
        }

        const payload: TusharePayload = {
            api_name: apiName,
            token,
            params,
            fields,
        };

        const response = await this.client.post<TushareResponse>(TUSHARE_CONFIG.BASE_URL, payload);
        if (response.data.code !== 0) {
            throw new Error(`Tushare API 错误: ${response.data.msg}`);
        }

        return mapFields(response.data.data);
    }

    private async fetchDaily(tsCode: string, range: { start_date: string; end_date: string }) {
        const data = await this.callApi('daily', { ts_code: tsCode, ...range }, DEFAULT_FIELDS.daily);
        return data.sort((a: any, b: any) => String(a.trade_date).localeCompare(String(b.trade_date)));
    }

    private async fetchDailyBasic(tsCode: string, tradeDate: string) {
        // Updated to query a range if tradeDate is empty, effectively getting latest
        const params: Record<string, any> = { ts_code: tsCode };
        if (tradeDate) {
            params.trade_date = tradeDate.replace(/-/g, '');
        } else {
            // If no specific date, use a range logic implicitly or default to latest available via external logic?
            // Actually tushare daily_basic without trade_date returns history, which we can sort.
            // But we need to limit it.
            const range = buildDateRange(5);
            params.start_date = range.start_date;
            params.end_date = range.end_date;
        }

        const data = await this.callApi('daily_basic', params, DEFAULT_FIELDS.dailyBasic);
        if (data && data.length > 0) {
            // Sort by date desc
            data.sort((a: any, b: any) => String(b.trade_date).localeCompare(String(a.trade_date)));
            return data[0];
        }
        return null;
    }

    private async fetchStockBasic(tsCode: string) {
        const data = await this.callApi('stock_basic', { ts_code: tsCode, list_status: 'L' }, DEFAULT_FIELDS.stockBasic);
        return data[0] || null;
    }

    private async fetchFinaIndicator(tsCode: string, range: { start_date: string; end_date: string }) {
        return this.callApi('fina_indicator', { ts_code: tsCode, ...range }, DEFAULT_FIELDS.finaIndicator);
    }

    private async fetchIncome(tsCode: string, range: { start_date: string; end_date: string }) {
        return this.callApi('income', { ts_code: tsCode, ...range }, DEFAULT_FIELDS.income);
    }

    private resolveToken(): string | undefined {
        const envToken = String(process.env.TUSHARE_TOKEN || '').trim();
        if (envToken) {
            this.token = envToken;
        }
        return this.token;
    }

    /**
     * 获取北向资金数据 (使用 moneyflow_hsgt 接口)
     */
    async getNorthFund(days: number): Promise<Array<{
        date: string;
        shConnect: number;
        szConnect: number;
        total: number;
        cumulative: number;
    }>> {
        const cacheKey = CacheAdapter.generateKey('northfund', SOURCE, days.toString());
        const cached = cache.get<Array<{ date: string; shConnect: number; szConnect: number; total: number; cumulative: number }>>(cacheKey);
        if (cached) return cached;

        const result = await rateLimiter.schedule(SOURCE, async () => {
            const range = buildDateRange(days);

            // 使用 moneyflow_hsgt 接口获取沪深港通资金流向
            const fields = 'trade_date,ggt_ss,ggt_sz,hgt,sgt,north_money,south_money';
            const data = await this.callApi('moneyflow_hsgt', {
                start_date: range.start_date,
                end_date: range.end_date,
            }, fields);

            if (!data || data.length === 0) {
                return [];
            }

            // 按日期排序
            const sorted = data.sort((a: any, b: any) =>
                String(a.trade_date).localeCompare(String(b.trade_date))
            );

            let cumulative = 0;
            return sorted.map(row => {
                // hgt = 沪股通, sgt = 深股通, north_money = 北向资金 (亿)
                const hgt = Number(row.hgt || 0);
                const sgt = Number(row.sgt || 0);
                const total = Number(row.north_money || (hgt + sgt));
                cumulative += total;

                return {
                    date: formatTradeDate(row.trade_date),
                    shConnect: hgt * 100000000,  // 转换为元
                    szConnect: sgt * 100000000,
                    total: total * 100000000,
                    cumulative: cumulative * 100000000,
                };
            }).slice(-days);
        });

        if (result.length > 0) {
            cache.set(cacheKey, result, CACHE_TTL.NORTH_FUND);
        }
        return result;
    }
}

export const tushareAdapter = new TushareAdapter();
